linux/run-nrtest.sh

```diff
@@ -6,85 +6,96 @@
+PROJ_DIR=${PWD}
 
-nrtest_execute_cmd="nrtest execute"
-test_app_path="apps/swmm-$3.json"
-tests="tests/examples tests/extran tests/routing tests/user"
-test_output_path="benchmark/swmm-$3"
 
-nrtest_compare_cmd="nrtest compare"
-ref_output_path="benchmark/swmm-${benchmark_ver}"
+# change current directory to test suite
+cd ${TEST_HOME}
 
-rtol_value=0.01
-atol_value=0.0
+# check if app config file exists
+if [[ ! -f "./apps/${PROJECT}-${SUT_BUILD_ID}.json" ]]
+then
+    mkdir -p "apps"
+    ${SCRIPT_HOME}/app-config.zsh "${PROJ_DIR}/${BUILD_HOME}/bin" \
+    ${SUT_BUILD_ID} ${SUT_VERSION} > "./apps/${PROJECT}-${SUT_BUILD_ID}.json"
+fi
+
+# build list of directories contaiing tests
+TESTS=$( find ./tests -mindepth 1 -type d -follow | paste -sd " " - )
 
-# change current directory to test_suite
-cd ${test_suite_path}
+# build nrtest execute command
+NRTEST_EXECUTE_CMD='nrtest execute'
+TEST_APP_PATH="./apps/${PROJECT}-${SUT_BUILD_ID}.json"
+TEST_OUTPUT_PATH="./benchmark/${PROJECT}-${SUT_BUILD_ID}"
 
-# clean test benchmark results
-rm -rf ${test_output_path}
+# build nrtest compare command
+NRTEST_COMPARE_CMD='nrtest compare'
+REF_OUTPUT_PATH="benchmark/${PROJECT}-${REF_BUILD_ID}"
+RTOL_VALUE='0.01'
+ATOL_VALUE='1.E-6'
 
-echo
-echo INFO: Creating test benchmark
-nrtest_command="${nrtest_execute_cmd} ${test_app_path} ${tests} -o ${test_output_path}"
-echo INFO: "$nrtest_command"
-return_value=$( $nrtest_command )
 
-if [ $1 = 'true' ]; then
-  echo
-  echo INFO: Comparing test and ref benchmarks
-  nrtest_command="${nrtest_compare_cmd} ${test_output_path} ${ref_output_path} --rtol ${rtol_value} --atol ${atol_value}"
-  echo INFO: "$nrtest_command"
-  return_value=$( $nrtest_command )
+# if present clean test benchmark results
+if [ -d "${TEST_OUTPUT_PATH}" ]; then
+    rm -rf "${TEST_OUTPUT_PATH}"
 fi
 
-return $return_value
-}
-
-
-print_usage() {
-   echo " "
-   echo "run-nrtest.sh - generates artifacts for SUT and performes benchmark comparison "
-   echo " "
-   echo "options:"
-   echo "-c,             don't compare SUT and benchmark artifacts"
-   echo "-t test_path    relative path to location where test suite is staged"
-   echo "-v version      version/build identifier"
-   echo " "
-}
-
-# Default option values
-compare='true'
-test_path='nrtestsuite'
-version='vXXX'
-
-while getopts ":ct:v:" flag; do
-  case "${flag}" in
-    c  ) compare='false' ;;
-    t  ) test_path="${OPTARG}" ;;
-    v  ) version=${OPTARG} ;;
-    \? ) print_usage
-         exit 1 ;;
-  esac
-done
-shift $(($OPTIND - 1))
+# perform nrtest execute
+echo "INFO: Creating SUT ${SUT_BUILD_ID} artifacts"
+NRTEST_COMMAND="${NRTEST_EXECUTE_CMD} ${TEST_APP_PATH} ${TESTS} -o ${TEST_OUTPUT_PATH}"
+eval ${NRTEST_COMMAND}
+
+# perform nrtest compare
+echo "INFO: Comparing SUT artifacts to REF ${REF_BUILD_ID}"
+NRTEST_COMMAND="${NRTEST_COMPARE_CMD} ${TEST_OUTPUT_PATH} ${REF_OUTPUT_PATH} --rtol ${RTOL_VALUE} --atol ${ATOL_VALUE}"
+eval ${NRTEST_COMMAND}
+
 
```

Comment:
It is useful to handle artifact upload here. Getting the benchmark files off the build worker when tests fail is useful for two reasons. 1) It allows you to study the files for errors, 2) if everything checks out you can use it to update the benchmark.

Reply:
@michaeltryby, I am a little confused here. Could you clarify? Are you saying that once the test passes/fails we want to be able to upload the artifacts back to (I am assuming) github right?

Reply:
@cbuahin Sorry about the confusion. My comment wasn't clear. 

I have found that the thing that makes artifact upload difficult is configuring the path on the build worker. Having to figure out the path to each artifact so the CI knows what to upload is a pain. So I tried something slightly different. I set up a designated location "upload" and copy artifacts I want uploaded into it. Much easier ...  Make sense? 

You can see it in action [here](https://github.com/michaeltryby/ci-tools/blob/dev/windows/run-nrtests.cmd).

Reply:
@michaeltryby, got it! That is actually a neat approach. Thanks for the clarification.